feat(platform-browser): enable HTTP request caching when using `provideClientHydration`

alan-agius4 opened the pull request:
This commit adds support by default for HTTP caching when using `provideClientHydration`. Users can opt-out of this behaviour by using the `withoutHttpTransferCache` feature.

```ts
import {
  bootstrapApplication,
  provideClientHydration,
  withNoHttpTransferCache,
} from '@angular/platform-browser';
// ...
bootstrapApplication(RootCmp, {
  providers: [provideClientHydration(withNoHttpTransferCache())]
});
```

alan-agius4 commented on packages/platform-browser/src/hydration.ts:
```diff
@@ -30,49 +36,36 @@ export interface HydrationFeature<FeatureKind extends HydrationFeatureKind> {
  * @developerPreview
  */
-export function withoutDomReuse(): NoDomReuseFeature {
+export function withoutDomReuse(): HydrationFeature<HydrationFeatureKind.NoDomReuseFeature> {
```
Are `HydrationFeature` and `HydrationFeatureKind` really needed to be part of the public API? In `@angular/common/http` I noticed they are but was wondering what they are needed.

AndrewKushnir replied:
I followed the typings that we have in the Router, but I'm ok with this option too 👍

alan-agius4 replied:
Should they still be exported?

alan-agius4 commented on packages/platform-browser/src/hydration.ts:
```diff
@@ -30,49 +36,36 @@ export interface HydrationFeature<FeatureKind extends HydrationFeatureKind> {
- * @publicApi
- * @developerPreview
- */
-export type NoDomReuseFeature = HydrationFeature<HydrationFeatureKind.NoDomReuseFeature>;
```
Removed the types to reduce the public API surface.

alan-agius4 commented:
**Caretaker note:** this requires a G3 patch to be updated. http://cl/521719612

AndrewKushnir commented on packages/platform-browser/src/hydration.ts:
```diff
@@ -102,19 +95,31 @@ export type HydrationFeatures = NoDomReuseFeature;
-export function provideClientHydration(...features: HydrationFeatures[]): EnvironmentProviders {
-  const shouldUseDomHydration =
-      !features.find(feature => feature.ɵkind === HydrationFeatureKind.NoDomReuseFeature);
+export function provideClientHydration(...features: HydrationFeature<HydrationFeatureKind>[]):
```
Since both features that we have here are non-tree-shakable anyway, should we consider an API like this instead?
```
provideClientHydration(withConfig({domReuse: false, httpCacheTransfer: false}));
```
This will still keep the door open for new features in a tree-shakable way, while simplifying the API a bit. WDYT?

// cc @jessicajaniuk

AndrewKushnir reviewed (commented):
LGTM, just a quick question on the API.

alan-agius4 committed:
feat(platform-browser): enable HTTP request caching when using `provideClientHydration`

This commit adds support by default for HTTP caching when using `provideClientHydration`. Users can opt-out of this behaviour by using the `withoutHttpTransferCache` feature.

```ts
import {
  bootstrapApplication,
  provideClientHydration,
  withNoHttpTransferCache,
} from '@angular/platform-browser';
// ...
bootstrapApplication(RootCmp, {
  providers: [provideClientHydration(withNoHttpTransferCache())]
});
```

AndrewKushnir reviewed (approved):
@alan-agius4 looks great! 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: fw-core, fw-http, fw-platform-server, public-api

AndrewKushnir commented:
**Caretaker notes**:

- this PR is ready for merge
- this change requires a G3 patch and BUILD file to be updated: http://cl/521719612 (just reposting this from [the comment above](https://github.com/angular/angular/pull/49699#issuecomment-1495786187) for visibility)
- the presubmit is "green" (only unrelated failures)

dylhunn commented:
This PR was merged into the repository by commit 81e7d15ef65b70c9734ebfd2c865e70d743263dc.
